Quarkus 3 quarkus.package.type=native-sources require docker/podman or native-image

### Describe the bug

Starting from version 3 of Quarkus, the behavior of the command './gradlew build '-Dquarkus.package.type=native-sources'' changed, as it now requires either Docker/Podman or native-image installed on the host. Up until version 2.16.0.Final, it did not require Docker or native-image to be installed.

### Expected behavior

It is expected not to require Docker/Podman or native-image, just as it worked in version 2.16.

### Actual behavior
```
 ./gradlew build -Dquarkus.package.type=native-sources                                                       ─╯

> Configure project :
WARNING: The task 'testCodeCoverageReport' of type 'JacocoReport' does not follow the naming convention "jacoco(TestTask)Report", where "TestTask" is the capitalized name of the test task covered by the report. We improvise by using 'jacocoLogCoverage13' as a fallback task name.
WARNING: The task 'integrationTestsCodeCoverageReport' of type 'JacocoReport' does not follow the naming convention "jacoco(TestTask)Report", where "TestTask" is the capitalized name of the test task covered by the report. We improvise by using 'jacocoLogCoverage14' as a fallback task name.
Cannot find the `native-image` in the GRAALVM_HOME, JAVA_HOME and System PATH. Install it using `gu install native-image` Attempting to fall back to container build.

> Task :quarkusBuild FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':quarkusBuild'.
> There was a failure while executing work items
   > A failure occurred while executing io.quarkus.gradle.tasks.worker.BuildWorker
      > io.quarkus.builder.BuildException: Build failure: Build failed due to errors
                [error]: Build step io.quarkus.deployment.pkg.steps.NativeImageBuildStep#resolveNativeImageBuildRunner threw an exception: java.lang.IllegalStateException: No container runtime was found. Make sure you have either Docker or Podman installed in your environment.
                at io.quarkus.runtime.util.ContainerRuntimeUtil.detectContainerRuntime(ContainerRuntimeUtil.java:99)
                at io.quarkus.runtime.util.ContainerRuntimeUtil.detectContainerRuntime(ContainerRuntimeUtil.java:41)
                at java.base/java.util.Optional.orElseGet(Optional.java:364)
                at io.quarkus.deployment.pkg.steps.NativeImageBuildContainerRunner.<init>(NativeImageBuildContainerRunner.java:31)
                at io.quarkus.deployment.pkg.steps.NativeImageBuildLocalContainerRunner.<init>(NativeImageBuildLocalContainerRunner.java:21)
                at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.resolveNativeImageBuildRunner(NativeImageBuildStep.java:332)
                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
                at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                at java.base/java.lang.reflect.Method.invoke(Method.java:568)
                at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
                at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
                at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
                at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
                at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
                at java.base/java.lang.Thread.run(Thread.java:833)
                at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```
### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

17

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.0.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

gardlew 8

### Additional information

_No response_

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java
@@ -329,7 +329,7 @@ public class NativeImageBuildStep {
     /**
      * Resolves the runner factory. Happens quite early, *before* the build.
      */
-    @BuildStep
+    @BuildStep(onlyIf = NativeBuild.class)
     public NativeImageRunnerBuildItem resolveNativeImageBuildRunner(NativeConfig nativeConfig) {
         boolean isExplicitContainerBuild = nativeConfig.containerBuild()
                 .orElse(nativeConfig.containerRuntime().isPresent() || nativeConfig.remoteContainerBuild());
@@ -351,6 +351,16 @@ public class NativeImageBuildStep {
             return new NativeImageRunnerBuildItem(new NativeImageBuildRemoteContainerRunner(nativeConfig));
         }
         return new NativeImageRunnerBuildItem(new NativeImageBuildLocalContainerRunner(nativeConfig));
+    }
+
+    /**
+     * Creates a dummy runner for native-sources builds. This allows the creation of native-source jars without
+     * requiring podman/docker or a local native-image installation.
+     */
+    @BuildStep(onlyIf = NativeSourcesBuild.class)
+    public NativeImageRunnerBuildItem dummyNativeImageBuildRunner(NativeConfig nativeConfig) {
+        boolean explicitContainerBuild = nativeConfig.isExplicitContainerBuild();
+        return new NativeImageRunnerBuildItem(new NoopNativeImageBuildRunner(explicitContainerBuild));
     }
 
     private void copyJarSourcesToLib(OutputTargetBuildItem outputTargetBuildItem,

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NoopNativeImageBuildRunner.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NoopNativeImageBuildRunner.java
@@ -1,0 +1,34 @@
+package io.quarkus.deployment.pkg.steps;
+
+import java.nio.file.Path;
+import java.util.List;
+
+public class NoopNativeImageBuildRunner extends NativeImageBuildRunner {
+
+    private static final String MESSAGE = "NoopNativeImageBuildRunner is not meant to be used to perform an actual build.";
+    private final boolean isContainer;
+
+    public NoopNativeImageBuildRunner(boolean isContainer) {
+        this.isContainer = isContainer;
+    }
+
+    @Override
+    public boolean isContainer() {
+        return isContainer;
+    }
+
+    @Override
+    protected String[] getGraalVMVersionCommand(List<String> args) {
+        throw new UnsupportedOperationException(MESSAGE);
+    }
+
+    @Override
+    protected String[] getBuildCommand(Path outputDir, List<String> args) {
+        throw new UnsupportedOperationException(MESSAGE);
+    }
+
+    @Override
+    protected void objcopy(Path outputDir, String... args) {
+        throw new UnsupportedOperationException(MESSAGE);
+    }
+}

--- a/docs/src/main/asciidoc/building-native-image.adoc
+++ b/docs/src/main/asciidoc/building-native-image.adoc
@@ -786,10 +786,11 @@ $ ./mvnw clean package -Dquarkus.package.type=native-sources -Dquarkus.native.co
 cd target/native-sources
 docker run \
   -it \
+  --user $(id -ur):$(id -gr) \
   --rm \
-  --v $(pwd):/work \# <1>
+  -v $(pwd):/work \# <1>
   -w /work \# <2>
-  --entrypoint bin/sh \
+  --entrypoint /bin/sh \
   $(cat native-builder.image) \# <3>
   -c "native-image $(cat native-image.args) -J-Xmx4g"# <4>
 ----

--- a/extensions/kafka-streams/deployment/src/main/java/io/quarkus/kafka/streams/deployment/KafkaStreamsProcessor.java
+++ b/extensions/kafka-streams/deployment/src/main/java/io/quarkus/kafka/streams/deployment/KafkaStreamsProcessor.java
@@ -32,6 +32,7 @@ import io.quarkus.deployment.builditem.nativeimage.NativeImageResourceBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.RuntimeReinitializedClassBuildItem;
 import io.quarkus.deployment.pkg.builditem.NativeImageRunnerBuildItem;
+import io.quarkus.deployment.pkg.steps.NativeOrNativeSourcesBuild;
 import io.quarkus.kafka.streams.runtime.KafkaStreamsProducer;
 import io.quarkus.kafka.streams.runtime.KafkaStreamsRecorder;
 import io.quarkus.kafka.streams.runtime.KafkaStreamsRuntimeConfig;
@@ -42,7 +43,7 @@ class KafkaStreamsProcessor {
 
     public static final String DEFAULT_PARTITION_GROUPER = "org.apache.kafka.streams.processor.DefaultPartitionGrouper";
 
-    @BuildStep
+    @BuildStep(onlyIf = NativeOrNativeSourcesBuild.class)
     void build(BuildProducer<FeatureBuildItem> feature,
             BuildProducer<ReflectiveClassBuildItem> reflectiveClasses,
             BuildProducer<JniRuntimeAccessBuildItem> jniRuntimeAccessibleClasses,

